pkg/arvo/neo/cod/std/src/fil/axal.txt

```diff
@@ -0,0 +1,338 @@
+# Axal guide
+
+`+axal` is the recursive tree type that Shrubbery uses to build the namespace.
+
+For userspace developers, axals are useful when you're writing `/con` files for your frontend and you want to work with the descendants of whichever shrub you're viewing in the frontend. (e.g. You're rendering a todo item and you want to render its sub-tasks, which are nodes beneath the main task in the namespace.)
+
+We work with axals using the `+of` core defined in `/sur/neo.hoon`, whose arms are documented below.
+
+[note on piths and paths, how `of:neo` is used and how to work with piths and paths, convert from path to pith, etc.]
+
+The `+axal` type itself looks like this.
+
+```
+++  axal
+  |$  [item]
+  [fil=(unit item) kid=(map iota $)]
+```
+
+One way to think of the Shrubbery namespace is as a filesystem, where files could be data, code, or something in between. So the `+axal` represents a directory of such files. The `fil` is the state of the file, which may or may not be null. The `kid` map stores everything below this file in the directory. The map's keys are the path segments immediately below this layer of the directory, and its values are the subtrees that correspond to those paths.
+
+***
+
+## `+anc`
+
+[??? - i see what this is doing just dk why]
+
+```
+++  anc
+  =|  res=(list pith)
+  =|  cur=pith
+  |=  pax=pith
+  ^-  (set pith)
+  ?~  pax
+    (~(gas in *(set pith)) res)
+  =?  res  ?=(^ fil.fat)
+    [cur res]
+  $(fat (~(got by kid.fat) i.pax), pax t.pax, cur (snoc cur i.pax))
+```
+
+???
+
+## `+anc-jab`
+
+[??? - apply gate to nodes?]
+
+```
+++  anc-jab
+  |*  [pax=pith fun=$-(* *)]
+  ^+  fat
+  ?~  pax
+    fat
+  =?  fil.fat  ?=(^ fil.fat)
+    `(fun u.fil.fat)
+  fat(kid (~(put by kid.fat) i.pax $(fat (~(got by kid.fat) i.pax), pax t.pax)))
+```
+
+???
+
+## `+del`
+
+Delete a node.
+
+```
+++  del
+  |=  pax=pith
+  ^+  fat
+  ?~  pax  [~ kid.fat]
+  =/  kid  (~(get by kid.fat) i.pax)
+  ?~  kid  fat
+  fat(kid (~(put by kid.fat) i.pax $(fat u.kid, pax t.pax)))
+```
+
+Returns an axal with the node at path `pax` removed.
+
+## `+dip`
+
+Descend to the axal at this path, [which does what]
+
+```
+++  dip
+  |=  pax=pith
+  ^+  fat
+  ?~  pax  fat
+  =/  kid  (~(get by kid.fat) i.pax)
+  ?~  kid  [~ ~]
+  $(fat u.kid, pax t.pax)
+```
+
+Returns an axal whose uppermost node is the node at the given path `pax`.
+
+## `+fit`
+
+Fetch file at the longest existing prefix of this path.
+
+```
+++  fit
+  |=  pax=pith
+  ^+  [pax fil.fat]
+  ?~  pax  [~ fil.fat]
+  =/  kid  (~(get by kid.fat) i.pax)
+  ?~  kid  [pax fil.fat]
+  =/  low  $(fat u.kid, pax t.pax)
+  ?~  +.low
+    [pax fil.fat]
+  low
+```
+
+Returns a `(pair pith *)` where `*` is the node at the given path `pax`.
+
+## `+gas`
+
+Merge a list of `(pair pith *)` with the axal.
+
+[??? that's what it looks like this should do but i'd expect a `|-` in here]
```

Comment:
yes that's correct. there actually should be a `|-` there lmao, I was wondering why the hell my gas:of calls weren't working correctly (see: +gas-epic and friends in app/neo)